### ISSUE TITLE
Running concurrent jobs on Jenkins with fastlane/scan puts CoreSimulatorService in a deadlock state

### ISSUE THREAD

**Issue by FabienLydoire** [fastlane/fastlane#14528]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
We rely on fastlane/scan to perform the tests on our Jenkins CI server.
- When running non concurrent jobs (1 Jenkins job at a time), we got no issue.
- When running concurrent jobs (Multiple fastlane/scan at the same time), jobs get blocked.

We rely on [fastlane-plugin-simctl](https://github.com/plu/simctl/tree/master/fastlane-plugin-simctl) to execute `xcrun simctl` commands.

We execute up to 12 concurrent Jenkins jobs.

_Most of the time_, the Jenkins jobs run fine.
_Sometimes_ the `xcrun simctl` commands are not responding at some point, so the jobs are stuck during simulator creation or deletion.

The process `com.apple.CoreSimulator.CoreSimulatorService` seems stuck in a deadlock.
We ran `sample` on the process and the sample analysis did exhibit :
> Dispatch Thread Soft Limit: 64 reached in 8117 of 8121 samples -- too many dispatch threads blocked in synchronous operations

Killing the simulators does not end the deadlock. We have to 
```killall -9 com.apple.CoreSimulator.CoreSimulatorService```

##### Command executed
```
private_lane :jenkins_test do |options|
	setup_jenkins(result_bundle: false)
	sh 'defaults write com.apple.iphonesimulator ConnectHardwareKeyboard 0'
	simctl(type: 'iPhone Xʀ',
		   block: lambda { |action, device|
		   action.scan(
					   skip_detect_devices: true,
					   destination: ["id=#{device.udid}"],
					   workspace: "#{Dir.pwd}/MyApp.xcworkspace",
					   scheme: options[:scheme],
					   xcargs: "-UseModernBuildSystem=NO",
					   clean: true,
					   code_coverage:true,
					   derived_data_path: '.',
					   output_directory: "./test_output/",
					   fail_build: false,
					   result_bundle: "TestResults" # To generate test reports
					   )
		   })
end
```

```
$ bundle exec fastlane jenkins_test scheme:"MyApp" configuration:"Debug"
```

### Environment

<details><summary>🚫 fastlane environment 🚫</summary>

### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 10.14.4                                                        |
| Ruby                        | 2.3.7                                                          |
| Bundler?                    | false                                                          |
| Git                         | git version 2.20.1 (Apple Git-117)                             |
| Installation Source         | /usr/local/bin/fastlane                                        |
| Host                        | Mac OS X 10.14.4 (18E226)                                      |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib |
| OpenSSL Version             | LibreSSL 2.6.5                                                 |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version               | 10.2                                                           |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE | en_US.UTF-8 | ✅ |
</details>

**Comment by FabienLydoire** [comment]
Hi @joshdholtz, that's the issue you wanted to see on Twitter.

**Comment by JoeAlamo** [comment]
@FabienLydoire Were you able to get to the bottom of this? I'm wanting to look into this approach as we currently have issues with multiple builds using same simulator (CoreData issues etc.), but I don't want to expend the effort just to run into this type of problem 😨

**Comment by FabienLydoire** [comment]
@JoeAlamo We reduced the number of concurrent jobs to 4. We **still get deadlocks** but less often.
We have an open Apple radar (rdar://49640668), and got unofficial reply that they are looking into it.
See https://twitter.com/FabienLydoire/status/1115890979553402881

**Comment by JoeAlamo** [comment]
Thank you for sharing that. Out of interest, what simctl commands are you running? Does that simctl action create a new simulator just for execution of that code block, or is it searching for the inputted device type and trying to find an existing simulator which matches it?

**Comment by FabienLydoire** [comment]
`simctl` is used to create a new simulator of the passed type (no reuse), run the `scan` command and then delete the simulator.
```
simctl(type: 'iPhone Xʀ',
		   block: lambda { |action, device|
		   	action.scan(
				skip_detect_devices: true,
				destination: ["id=#{device.udid}"],
				workspace: "#{Dir.pwd}/MyApp.xcworkspace",
				scheme: options[:scheme],
					   clean: true,
					   code_coverage:true,
					   derived_data_path: '.',
					   output_directory: "./test_output#{options[:report_suffix]}/",
					   fail_build: false,
					   result_bundle: "TestResults" # To generate test reports
				 )
		   })
```

**Comment by JoeAlamo** [comment]
Thanks for that. If you end up hearing back from Apple or implement workarounds I would greatly appreciate it if you could update back here 🙏

I may be trying this same approach , on a couple of build runners that perform concurrent jobs (using Gitlab but should make no difference) across multiple distinct projects. If I do so I'll report back if successful or not.

Lack of isolation in general on MacOS has caused us a lot of problems, when originally coming from Docker based CI/CD. It would be really unfortunate to have to resort to VM-per-build type setup due to resource consumption & speed reduction, but it may be required for stable builds 😞

**Comment by FabienLydoire** [comment]
@JoeAlamo 
See https://twitter.com/FabienLydoire/status/1114159992314236928
@lyndsey-ferguson is running 1 physical macOS computer per simulator.

**Comment by lyndsey-ferguson** [comment]
A couple of thoughts:
1. the act of creating a simulator can help fix issues with a particular instance of a sim, but this is very slow. I would recommend that if you know the number of simulators you need, to pre-create a pool and use those.
1. creating a simulator when the CoreSimulatorService will be a waste of time because the service that connects to the simulator is broken. You'll have unload that launch agent and relaunch it to fix the issue.
1. Yes, I have dedicated one machine to one Jenkins job so as to avoid issues with `:derived_data` paths being shared between jobs (which can be fixed by providing unique paths), and the issue with multiple simulators. I believe that if you keep your numbers of executors on a machine to a low enough level you should not encounter too many deadlocks.

**Comment by JoeAlamo** [comment]
👍 Thanks for sharing your experiences Lyndsey. 

How long was it taking for you to create a simulator?

Usually the types of issues that we experience with simulators is often CoreData related, where build 1 on Branch A will run one set of data migrations, and then build 2 on Branch B will try to run a different set, experience conflicts and break. This requires erasing the app from the simulator to fix it.

I've thought about specifying the 'uninstall app before starting Scan' option, but I'm unsure how that would affect other test jobs running at the same time. For example, would it wait until tests were finished running on the simulator before uninstalling app?

**Comment by lyndsey-ferguson** [comment]
Simulator creation was roughly a second. For me, that is important as I am creating a pool of simulators to be used for parallel UI testing. To reduce connection problems with the `testmanagerd` that the CoreSimulator service manages, I find that if I create and boot them in a batch, the problems do not happen unless I ask for more than 6 parallel runs of `xcodebuild test`.

With regards to the conflicts you wrote about, I saw that too if the `derived data` path and the `build log paths` where the same. Once I started randomly generating those, after copying the `.xctestrun` files and accompanying data files to those folders, the conflicts stopped.

With regards to uninstalling, if your job is the _only_ job using that particular simulator, I cannot imagine that other jobs would have a problem. If you were using the same simulator, but testing with different simulators, you may be okay, but I could imagine that Apple hasn't anticipated this and you could have problems. I would just use a different sim for different jobs.

**Comment by JoeAlamo** [comment]
Interesting findings 🤔 thanks again, those details will really help me when implementing and tuning my setup 🙌 Sorry for partially hijacking your issue Fabien!

**Comment by FabienLydoire** [comment]
@JoeAlamo No problem. That's part of the same topic.
